src/bootstrap/builder.rs

```diff
@@ -968,22 +968,25 @@ impl<'a> Builder<'a> {
             cargo.env("RUSTDOC_LIBDIR", self.rustc_libdir(compiler));
         }
 
-        if mode.is_tool() {
-            // Tools like cargo and rls don't get debuginfo by default right now, but this can be
-            // enabled in the config.  Adding debuginfo makes them several times larger.
-            if self.config.rust_debuginfo_tools {
-                cargo.env("RUSTC_DEBUGINFO", self.config.rust_debuginfo.to_string());
-                cargo.env(
-                    "RUSTC_DEBUGINFO_LINES",
-                    self.config.rust_debuginfo_lines.to_string(),
-                );
-            }
-        } else {
-            cargo.env("RUSTC_DEBUGINFO", self.config.rust_debuginfo.to_string());
-            cargo.env(
-                "RUSTC_DEBUGINFO_LINES",
-                self.config.rust_debuginfo_lines.to_string(),
-            );
+        enum DebuginfoPartition {
+            Rustc,
+            Std,
+            Tools,
+        }
+        let debuginfo_partition = match mode {
+            Mode::Rustc | Mode::Codegen => DebuginfoPartition::Rustc,
+            Mode::Std | Mode::Test => DebuginfoPartition::Std,
+            Mode::ToolBootstrap | Mode::ToolStd |
+            Mode::ToolTest | Mode::ToolRustc => DebuginfoPartition::Tools,
+        };
+        let debuginfo_level = match debuginfo_partition {
+            DebuginfoPartition::Rustc => self.config.rust_debuginfo_level_rustc,
+            DebuginfoPartition::Std => self.config.rust_debuginfo_level_std,
+            DebuginfoPartition::Tools => self.config.rust_debuginfo_level_tools,
+        };
+        cargo.env("RUSTC_DEBUGINFO_LEVEL", debuginfo_level.to_string());
```

Comment:
re:rustdoc I would re-add the code to tool.rs that we had before, just set RUSTC_DEBUGINFO_LEVEL to the rustc level there.

Reply:
I ended up removing the exception from rustdoc.
If someone wants to debug rustdoc, they are unlikely to build "large" tools like clippy/rustfmt/etc simultaneously (they are not even included into the default build), so enabling symbols for all tools shouldn't be especially problematic.